testing/testPlan.md

```diff
@@ -0,0 +1,172 @@
+        self.assertTrue('FOO'.isupper())
+        self.assertFalse('Foo'.isupper())
+
+    def test_split(self):
+        s = 'hello world'
+        self.assertEqual(s.split(), ['hello', 'world'])
+        # check that s.split fails when the separator is not a string
+        with self.assertRaises(TypeError):
+            s.split(2)
+
+if __name__ == '__main__':
+    unittest.main()
+```
+
+The first test `test_upper()` demonstrates how simple it is to create a unit 
+test. This test requires no setup or teardown, only the UUT which is the 
+`.upper` function of the string class. `test_split()` is a test that shows a 
+setup, the `s = 'hello world'`, and also shows a demonstration of negative 
+testing. Negative testing is when we expect the UUT to NOT do something. An 
+example of this is seen in the above code where we expect that `s.split()` to 
+fail when the parameter to `s.split()` is the int 2, and not a string.
+
+The following is an example using a function `TraceFileParser.open` from our 
+from our codebase.
+
+```
+def open(self) -> bool: 
+    """Open the trace file""" 
+    try: 
+        self.fd = open(self.filename, 'r', encoding="utf-8") 
+        return True
+    except FileNotFoundError:    
+        print(f"[ERROR] - could not find trace file '{self.filename}'") 
+        return False
+    except PermissionError: 
+        print(f"[ERROR] - Permission denied accessing file: '{self.filename}'")
+        return False
+    except Exception as e: 
+        print(f"[ERROR] - Opening file: {e}")
+        return False
+```
+
+A unit test for this function could looks something like this:
+
+```
+class traceFileParser_Open(unittest.TestCase)
+
+    def test_openFile(self):
+        parser = TraceFileParser()
+        self.assertTrue(parser.open("filename"))
+    
+    def test_FileNotFoundError:
+        parser = TraceFileParser()
+        self.assertFalse(parser.open("bad_filename"))
+
+    def test_FileNotFoundError
+        parser = TraceFileParser()
+        self.assertFalse(parser.open("file_with_bad_permissions))
+
+    def test_FileNotFoundError
+        parser = TraceFileParser()
+        self.assertFalse(parser.open("file_with_exception"))
+```
+
+Note, this would be a good function to refactor for returning the error instead
+of immediately printing it. Then we can test the error message as well.
+
+[Link to more unit test examples](https://github.com/mdhardenburgh/matrixLibrary/blob/main/unitTest/matrixTest.cpp).
+These examples are in C++, but you can see the 2 of the 3 parts of a unit test: 
+setup and verification. These examples did not need the optional teardown. 
+
+#### Note on functions that do not return a value
+These are a bit harder to test. In my experience, the function is modified to
+return a value. However, this may not be desirable in some cases and debug 
+prints could be parsed by the unit test, if the former solution is completely 
+unworkable. 
+
+#### Note on naming convention
```

Comment:
For our case using `unittest`, will all of our test functions in a file need to be prepended with `test_`? That is, anything that should run as a test, versus something in that file that might be used for setup or teardown (or other utility)?

Reply:
The convention at Tektronix is to prepend our tests with `test_` so I went with that. Anything that's used for setup or teardown we dont put `test_` in front of.

Reply:
Ok, I was just checking whether `unittest` has any convention or requirement we'd need to follow, and whether it makes sense to address that in this section.